### PR TITLE
fix(gradle-plugin): Kubernetes Gradle Build tasks don't account for image model configuration skip field (#1197)

### LINKED ISSUE #1197: [Kubernetes Gradle Plugin] Image build is not skipped if `build` > `skip` is set to `true`
### Description
While investigating https://github.com/eclipse/jkube/issues/1195, I noticed in gradle plugins we're not considering `skip` field in build configuration. 

If I add a `<skip>true</skip>` field in image build configuration like this:

```xml
      <plugin>
        <groupId>org.eclipse.jkube</groupId>
        <artifactId>kubernetes-maven-plugin</artifactId>
        <version>${jkube.version}</version>
        <configuration>
          <images>
            <image>
              <name>${project.groupId}/${project.artifactId}</name>
              <build>
                <from>quay.io/jkube/jkube-java-binary-s2i:${jkube.image.version}</from>
                <skip>true</skip>
```
I can see image build is skipped on executing `k8s:build`:
```
[INFO] Building Eclipse JKube :: Quickstarts :: Maven :: Micronaut customized Image 1.5.1
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- kubernetes-maven-plugin:1.5.1:build (default-cli) @ micronaut-customized-image ---
[INFO] k8s: Running in Kubernetes mode
[INFO] k8s: Building Docker image in Kubernetes mode
[INFO] k8s: [org.eclipse.jkube.quickstarts.maven/micronaut-customized-image:latest] : Skipped building
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.163 s
[INFO] Finished at: 2022-01-06T22:14:19+05:30
[INFO] ------------------------------------------------------------------------
```

However, If I add `skip=true` field in quickstarts/gradle/micronaut-customized-image, image building is not skipped:

```groovy
kubernetes {
    images {
        image {
            name = "%g/%a:${project.version}"
            build {
                skip = true
               // ...
            }
        }
    }
```

We need to add filter images like done in AbstractDockerMojo in KubernetesBuildTask also:
We can filter out image configurations here which don't have `skip` enabled like done for build goal:
https://github.com/eclipse/jkube/blob/b66213473b17a4d4c25096fc02055ce33197797b/kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/AbstractDockerMojo.java#L485-L489
### Info

* Eclipse JKube version : 1.5.1

* Kubernetes / Red Hat OpenShift setup and version : N/A

* If it's a bug, how to reproduce : go to quickstarts/gradle/micronaut-customized-image and do `gradle k8sBuild` after setting `skip = true` in `build { .. }` section

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

Fix #1197

Kubernetes/OpenShift Gradle Build tasks now take `image` > `build` > `skip` field in consideration while building images

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

## Type of change

 - [X] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [X] I have implemented unit tests to cover my changes
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [X] I tested my code in Kubernetes
 - [X] I tested my code in OpenShift

### CODE REVIEW

File: jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/BuildService.java [hunk 1]
```diff
@@ -31,11 +32,20 @@
...
-    void build(ImageConfiguration imageConfig) throws JKubeServiceException;
+    void build(Collection<ImageConfiguration> imageConfigs) throws JKubeServiceException;
+
+    /**
+     * Build the given image using specified configuration
+     *
+     * @param imageConfiguration single {@link ImageConfiguration} to build
+     * @throws JKubeServiceException in case of any error while building image
+     */
+    void build(ImageConfiguration imageConfiguration) throws JKubeServiceException;
```
Comment (manusa) [inline jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/BuildService.java:48]: Why did we keep the legacy method?

Comment (rohanKanojia) [reply jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/BuildService.java:48]: I've shifted image build skip logic to BuildService interface. this method gets called when image build is not skipped

Comment (manusa) [reply jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/BuildService.java:48]: After your update, I still see that we keep the two methods List and single image.

Is there a specific reason to keep both (at least in the public interface)??

File: jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/AbstractImageBuildService.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  protected AbstractImageBuildService(JKubeServiceHub jKubeServiceHub) {
+    this.jKubeServiceHub = jKubeServiceHub;
+  }
+
+  @Override
+  public void build(Collection<ImageConfiguration> imageConfigurations) throws JKubeServiceException {
+    buildImages(imageConfigurations);
+  }
+
+  protected void buildImages(Collection<ImageConfiguration> imageConfigurations) throws JKubeServiceException {
```
Comment (manusa) [inline jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/AbstractImageBuildService.java:32]: why not just a single method?

File: jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/BuildService.java (hunk 1 above)
Comment (manusa) [reply jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/BuildService.java:47]: https://github.com/rohanKanojia/jkube/pull/60

File: gradle-plugin/openshift/src/test/java/org/eclipse/jkube/gradle/plugin/task/OpenShiftBuildTaskTest.java [hunk 3]
```diff
@@ -41,12 +41,12 @@
   public TaskEnvironment taskEnvironment = new TaskEnvironment();
 
   private MockedConstruction<OpenshiftBuildService> openshiftBuildServiceMockedConstruction;
+  private final OpenShiftExtension extension = new TestOpenShiftExtension();
```
Comment (manusa) [inline gradle-plugin/openshift/src/test/java/org/eclipse/jkube/gradle/plugin/task/OpenShiftBuildTaskTest.java:44]: NOT OK, either revert this change, or initialize the variable in the setUp method.

**Review by manusa — APPROVED** [review]
LGTM, thx!

File: gradle-plugin/kubernetes/src/main/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTask.java [hunk 4]
```diff
@@ -57,14 +57,12 @@ public void run() {
...
-            getBuildTimestamp(null, null, kubernetesExtension.javaProject.getBuildDirectory().getAbsolutePath(),
-                DOCKER_BUILD_TIMESTAMP));
-        jKubeServiceHub.getBuildService().build(imageConfig);
-      }
+      storeTimestamp(
+          getBuildTimestampFile(kubernetesExtension.javaProject.getBuildDirectory().getAbsolutePath(),
+              DOCKER_BUILD_TIMESTAMP),
+          getBuildTimestamp(null, null, kubernetesExtension.javaProject.getBuildDirectory().getAbsolutePath(),
+              DOCKER_BUILD_TIMESTAMP));
+      jKubeServiceHub.getBuildService().build(resolvedImages.toArray(new ImageConfiguration[0]));
```
Comment (sunix) [inline gradle-plugin/kubernetes/src/main/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTask.java:65]: what is the reason of getting only the first image configuration?

Comment (sunix) [reply gradle-plugin/kubernetes/src/main/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTask.java:65]: I feel like this is fragile, could `resolvedImages` be empty?

Comment (manusa) [reply gradle-plugin/kubernetes/src/main/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTask.java:65]: This is just a file with the timestamp which is reused in all images.
After many refactors this has been reduced to what you see now. It belongs somewhere else, but that should be handled in an additional refactor.

Comment (manusa) [reply gradle-plugin/kubernetes/src/main/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTask.java:65]: > I feel like this is fragile, could resolvedImages be empty?

They can be empty, but can't be null

**Review by sunix — APPROVED** [review]

**Pull request merged**